### ISSUE TITLE
aarch64 flatpak builds: how to proceed.

### ISSUE THREAD

**Issue by leamas** [Rasbats/shipdriver_pi#153]
We need to start building aarch64/arm64 flatpak builds. This is a bit complicated since flatpak-builder, the tool used to build these plugins cannot run in an unprivileged container. This makes it or more or less impossible to use our current providers like CircleCI or drone.io. 

There is a lot of things happening in the aarch64 space, I'm sure alternatives  will come up, see https://github.com/OpenCPN/OpenCPN/issues/2178. But for now the best alternative I can see is using Github Actions (GA) to build the plugins. GA hasn't any support for aarch64, but it has a concept called self-hosted runners. Basically this means that while GA keeps track of administration, GUI, etc. the actual build is running on customer computer(s).

I have a preliminary setup which builds flatpak plugins for x86_64 on my general-purpose development machine, and aarch64 builds on a Rpi 4. it seems to "work", at least occasionally.

The main problem is security. Anything built here runs with too much privileges. I just cannot allow anyone to make a PR which executes arbitrary code on my machines. This means that the repo used to build the flatpak plugins must either be private or blocked from PR:s. Either solution is fine for me, but doesn't work for you.

The basic idea I have is using a github "organization" (basically, a group) and have repos in there which mirrors the "real" ones which still have public access. So, the workflow for you would be something like:

  - modify regular shipdriver_pi repo
  - commit and push -> all plugins besides flatpak built
  - import changes from regular shipdriver_pi  to a private mirror in the new group
  - push changes ->  flatpak plugins built.

I understand that this somewhat painful, and I'm open to other ideas as long as access to my machines is restricted (of course, I have absolutely no problem giving you access).

I will invite you to my new group to get things moving. Bot nothing, absolutely nothing is set in stone here.

Thoughts?

**Comment by leamas** [comment]
Hi again (sorry for too much text...),

I have moved the regular flatpak builder to a VM. This means  nothing runs in my regular development box, much less security problems. If you have time, I'd appreciate if you could do some tests:

  - Accept the invitation to the leamas-ga group (this is the group which has access to the flatpak builders).
  - Create an  empty repository in leamas-ga called _leamas-ga/shipdriver-mike_ using Github GUI
  - Using cmd, go to the shipdriver_pi parent directory. Here, create  a bare clone named shipdriver_pi.git and  ush this as a mirror copy  to the leamas-ga/shipdriver-mike  repo in Github
  
         > git clone --bare shipdriver_pi
         > cd shipdriver_pi.git
         > git remote add  ga <insert leamas-ga/shipdriver-mike git repo url from Github UI>
         > git push --mirror ga master:master
  - Now you have a mirror of Rasbats/shipdriver_pi in leamas-ga/shipdriver-mike.  Clone this as usual using the git url. This creates a local directory shipdriver-mike. Then

        > cd shipdriver-mike
        > git remote add upstream https://github.com/Rasbats/shipdriver_pi.git
    
  - This is the end of the initial setup. When you want to actually build, fetch the changes in Rasbats/shipdriver_pi and push them to leamas-ga/shipdriver-mike:
    
        > cd shipdriver-mike
        > git remote update upstream
        > git pull upstream master
        > git push origin master:master

This will actually build all plugins, we need to look into that. But for now, the important thing is to get things running.

**Comment by Rasbats** [comment]
Still digesting. I will action the git stuff tomorrow.

I prefer having a non-virtual system to test stuff. Need an excuse for getting an RPi 4. This could be it. How much work does it take to set up flatpak-builder? Can local builds be made this way with an arm64 image? Need to do some research!

Support going forward may be better with Raspberry Pi OS arm64 (Debian Buster). Is this likely to be the case or is there another distro to start playing with?

**Comment by leamas** [comment]
> How much work does it take to set up flatpak-builder?

It's simple, just a package to install. The build script actually does it, it's just a few lines.

> Can local builds be made this way with an arm64 image?

Yes: https://github.com/Rasbats/managed_plugins/wiki/Local-Build

> Support going forward may be better with Raspberry Pi OS arm64 (Debian
> Buster). Is this likely to be the case or is there another distro to
> start playing with

Go for the standard build OS:es instead. If the focus is  flatpak, go for Fedora. Otherwise Ubuntu. In any case, use recent version (F33, Focal). Current build script uses Fedora-33

Using an Rpi-4 you could also do the same as I have done: create your own aarch64 builder which connects into the Github Actions framework. The official specs says that you need 8GB of memory for that. Perhaps overkill. 

OTOH, my machine is reasonably stable in a 19" rack. Perhaps better to use your as test box and my for builds.

**Comment by Rasbats** [comment]
Still digesting. I will action the git stuff tomorrow.

I prefer having a non-virtual system to test stuff. Need an excuse for getting an RPi 4. This could be it. How much work does it take to set up flatpak-builder? Can local builds be made this way with an arm64 image? Need to do some research!

Support going forward may be better with Raspberry Pi OS arm64 (Debian Buster). Is this likely to be the case or is there another distro to start playing with?

**Comment by Rasbats** [comment]
Do you need to turn off master branch protection in "Settings" -> "Protected branches"?
Getting this:
`error: --mirror can't be combined with refspecs`
When I try to do:
 `git remote add  ga`

**Comment by leamas** [comment]
This is better explained in https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/duplicating-a-repository

**Comment by Rasbats** [comment]
Thanks. That got the initial setup done.
`git clone --bare shipdriver_pi`
Should have been:
`git clone --bare https://github.com/Rasbats/shipdriver_pi.git`

**Comment by Rasbats** [comment]
Pushed a minor change (# in CMakeLists.txt) to [https://github.com/leamas-ga/shipdriver-mike](https://github.com/leamas-ga/shipdriver-mike).
That worked ok. How to make the build? And where are the plugins created? Is another Cloudsmith repo needed?

**Comment by leamas** [comment]
Sorry  I didn't get it right. You are the subject here, sort of.

Now, next step would be that I make a PR. I need to prepare that, we should not import my current, embarrassing history. Stay tuned.

While on it, I can see two workflows. The first alternative would be to build current plugins as today, and just the flatpak plugins in shipdriver-mike. The second would be to build everything in shipdriver-mike. 

Thoughts?

**Comment by Rasbats** [comment]
Would prefer to keep the the flatpak plugins separate in shipdriver-mike. And just the aarch64 ones?

**Comment by leamas** [comment]
As it is for now, both flatpak plugins are built on Fedora and thus on shipdriver-mike.  It is certainly possible to move the regular x86_64 builds to Ubuntu which would make it possible to build the together with the others. Can we wait with this now, here are too many loose ends? And build both flatpak plugins in shipdriver-mike?

I need to do some work so we can update the flatpak runtime as well. The current version 18.04 is on it's way out, and we should preferably not use it on the new aarch 64 platform. But we should first verify that the plugin works...

**Comment by leamas** [comment]
Building just the flatpak plugins would be about merging my PR and, possibly, enable Actions for shipdriver-mike. The latter is done in the github gui for shipdriver-mike, in Settings  | Actions.

**Comment by Rasbats** [comment]
> And build both flatpak plugins in shipdriver-mike?

That's fine.

**Comment by Rasbats** [comment]
>  update the environment variables for Actions (GA)

Done.

**Comment by leamas** [comment]
> How to make the build? 

  - Set up the environment, the usual stuff in shipdriver-mike, Settings | Secrets
  - Enable Github Actions in  shipdriver-mike, Settings  | Actions
  - Merge the PR #154

> And where are the plugins created? 

Everything is as usual, it's using the standard upload script  pushing things to cloudsmith.

> Is another Cloudsmith repo needed?

No

**Comment by Rasbats** [comment]
Ok... good work. That went well. Just small errors, I hope:
![build_error_circleci](https://user-images.githubusercontent.com/2537240/109648139-2bf5dd00-7b52-11eb-9f47-009201c91c85.jpg)

**Comment by leamas** [comment]
Probably small, but busy right now. BBL

**Comment by leamas** [comment]
Fixed in #155

**Comment by Rasbats** [comment]
Made some new builds. Looks good... no errors. Impressive build speeds.

**Comment by leamas** [comment]
Looks like we have achieved the "works occasionally" state then... 

Next step is to update the aarch64 plugins to use the updated runtime, which means building against opencpn in the beta repo. More work, I'm on it. Thanks for your efforts!

**Comment by leamas** [comment]
Also, note to me:  need to make a write-up in the wiki.

**Comment by leamas** [comment]
Did a write-up: https://github.com/Rasbats/managed_plugins/wiki/Github-Actions-(BETA,-not-REVIEWED,-BEWARE!). If/when you have some time I would appreciate if you could review .

**Comment by leamas** [comment]
Sorry, seems we had some mid-air collisions on the wiki. I'll keep my  hands off.

**Comment by Rasbats** [comment]
Just added a line about cloning (Item 4). Nothing more to add at the moment.

**Comment by leamas** [comment]
Thanks for reviewing with fresh eyes!

**Comment by Rasbats** [comment]
> While on it, I can see two workflows. The first alternative would be to build current plugins as today, and just the flatpak plugins in shipdriver-mike. The second would be to build everything in shipdriver-mike.

Having seen the workflow I think the second option would be easier to maintain. GA looks good but at the moment includes a single point of failure - your machine. I think we need CircleCI supporting Arm64. (Plenty of votes for this)

**Comment by leamas** [comment]
> Having seen the workflow I think the second option would be easier to maintain

Yes... I have had similar thoughts. I'll update the wiki in this direction, later this week. There are no code changes involved.

> GA looks good but at the moment includes a single point of failure - your machine. I think we need CircleCI supporting Arm64. (Plenty of votes for this)

Indeed.  I hesitated before entering this path for this reason, and for just not being comfortable with having this at home. I only see current solution as a transition phase until we can use CircleCI (or something else, the arm universe is expanding). There is a catch: we need real VM:s, not just Docker instances like drone. 

BTW: Setting up a builder like I did is *really* simple, Google shines when it comes to designing such services.

**Comment by leamas** [comment]
> > Having seen the workflow I think the second option would be easier to maintain

> Yes... I have had similar thoughts. I'll update the wiki in this direction, later this week. There is no code changes involved.

... but it doesn't work since CircleCi doesn't build a private  repository on the open-source plan. It's clumsy, but we need to have it this way where the public repo builds everything but the experimental, flatpak plugins on x86_64/aarch64/20.08 which are built in the private clone.

Let's hope CircleCI comes up with something which can replace this Real Soon ™.

**Comment by Rasbats** [comment]
Results of trying to run the plugin on aarch64:

[fedora_pi_64bit_opencpn.log.txt](https://github.com/Rasbats/shipdriver_pi/files/6124118/fedora_pi_64bit_opencpn.log.txt)

OpenCPN starts and runs fine.

**Comment by leamas** [comment]
no... not that one, I mean opencpn.log. It is in `~/.var/app/org.opencpn.OpenCPN/config/opencpn/opencpn.log`

**Comment by Rasbats** [comment]
Sorry... realised my mistake:

[opencpn.log](https://github.com/Rasbats/shipdriver_pi/files/6124235/opencpn.log)

**Comment by leamas** [comment]
OK, this is about /home/mikerossiter/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so  What does 

    ldd /home/mikerossiter/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so

say?

**Comment by leamas** [comment]
Another datapoint. Building and installing from current shipdriver tip at b5e3d94, I can import the plugin without problems. This is on Fedora (F33).

**Comment by leamas** [comment]
Hold... ldd must be run from within the sandbox to make sense. Instead of a plain ldd, please do

      $  flatpak run --command=bash --devel  org.opencpn.OpenCPN
      [📦 org.opencpn.OpenCPN build]$ ldd /home/mike*/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so
 

... and post the results.

**Comment by Rasbats** [comment]
`'not a dynamic executable'`

Tested with one of the built in plugins and ldd works as it should.

**Comment by leamas** [comment]
Oops... that looks really bad. what does `file /home/mike*/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so`say?

**Comment by Rasbats** [comment]
file /home/mike*/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so
/home/mikerossiter/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=f1441bc0c48b28cb5f59c5616b6524bbfc7ea47b, with debug_info, not stripped
[mikerossiter@localhost ~]$ 

G'Night

**Comment by Rasbats** [comment]
> $  flatpak run --command=bash --devel  org.opencpn.OpenCPN
  [📦 org.opencpn.OpenCPN build]$ ldd /home/mike*/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so

Missed that bit... tomorrow.

**Comment by leamas** [comment]
Start by attaching the complete .tar.gz tarball here so I can test the same thing.

**Comment by Rasbats** [comment]
[ShipDriver-2.6.3.0+2103051212.7f3cb41_flatpak-20.08-aarch64.tar.gz](https://github.com/Rasbats/shipdriver_pi/files/6128477/ShipDriver-2.6.3.0%2B2103051212.7f3cb41_flatpak-20.08-aarch64.tar.gz)

**Comment by leamas** [comment]
hm... it might be about aarch64 vs arm64. cmake/Getarch reports aarch 64 on Fedora (the build platform) so the plugin's arch is 'aarch64'. However, your test box is Debian based and uses arm64. 

Can you try attached tarball, rebuilt with arch == 'arm64'?

[ShipDriver-2.6.3.0+2103120914.7f3cb41_flatpak-18.08-arm64.tar.gz](https://github.com/Rasbats/shipdriver_pi/files/6128882/ShipDriver-2.6.3.0%2B2103120914.7f3cb41_flatpak-18.08-arm64.tar.gz)

**Comment by Rasbats** [comment]
file /home/mike*/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so
/home/mikerossiter/.var/app/org.opencpn.OpenCPN/lib/opencpn/libShipDriver_pi.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=f1441bc0c48b28cb5f59c5616b6524bbfc7ea47b, with debug_info, not stripped
[mikerossiter@localhost ~]$ 

G'Night

**Comment by leamas** [comment]
OK, but my idea was that you should try to import it into OpenCPN ;) (I have an aarch64 machine available, but it's headless)

**Comment by Rasbats** [comment]
> Can you try attached tarball, rebuilt with arch == 'arm64'?

'Incompatible plugin detected'

> However, your test box is Debian based and uses arm64.

Fedora 33 (Workstation Edition) 64bit running on Raspberry Pi 4B. Checked the OpenCPN version: aarch64 flatpak. Works fine. Built-in plugins work fine.

Everything is ARM aarch64 when I use the Cloudsmith aarch64 ShipDriver tarball????

**Comment by Rasbats** [comment]
> Everything is ARM aarch64 when I use the Cloudsmith aarch64 ShipDriver tarball????

Just to confirm: the tarball, rebuilt with arch == 'arm64' gave the incompatible message when imported.

**Comment by leamas** [comment]
OK, I misunderstood. Sorry for the mess :(

Next step would be the full opencpn.log log after starting from the command line in debug mode, like 'opencpn -l debug`. That should reveal why it doesn't accept the plugin.

Sigh. Running out of time for now.

**Comment by Rasbats** [comment]
> Can you try attached tarball, rebuilt with arch == 'arm64'?

'Incompatible plugin detected'

> However, your test box is Debian based and uses arm64.

Fedora 33 (Workstation Edition) 64bit running on Raspberry Pi 4B. Checked the OpenCPN version: aarch64 flatpak. Works fine. Built-in plugins work fine.

Everything is ARM aarch64 when I use the Cloudsmith aarch64 ShipDriver tarball????

**Comment by Rasbats** [comment]
Actually: 
`/usr/bin/flatpak run --branch=stable --arch=aarch64 --command=opencpn.sh org.opencpn.OpenCPN -l debug`
Interesting terminal output:

[fedora_pi_64bit_running_opencpn_debug.txt](https://github.com/Rasbats/shipdriver_pi/files/6130819/fedora_pi_64bit_running_opencpn_debug.txt)

[opencpn.log](https://github.com/Rasbats/shipdriver_pi/files/6130921/opencpn.log)

**Comment by leamas** [comment]
Hm...   a quick scan through that log shows not a single hit on 'flatpak-a' i. e., here is no plugin with a flatpak-aarch64 or flatpak-arm64. Have you actually imported the plugin in this log?

**Comment by Rasbats** [comment]
Both came up as 'incompatible plugin detected'.

**Comment by leamas** [comment]
Still confused, on a higher level... finding things like this in the log:

15:12:17.636   DEBUG PluginHandler.cpp:234 Plugin compatibility check1: Climatology  OS:flatpak-x86_64  Plugin:flatpak-x86_64
15:12:17.636   DEBUG PluginHandler.cpp:291 Plugin compatibility checkFinal ACCEPTED: Climatology  PluginOS:flatpak-x86_64  PluginVersion: 18.08


From the log, one could think that this was from a flatpak-x86_64 machine. What am I missing?

**Comment by Rasbats** [comment]
> 15:12:17.636 DEBUG PluginHandler.cpp:291 Plugin compatibility checkFinal ACCEPTED: Climatology PluginOS:flatpak-x86_64 

Quite... the plugins in the managed list will install. They then appear in the list with the jig

**Comment by leamas** [comment]
All this boils down to that despite we are running the flatpak'ed opencpn on a arm-64 host it still thinks it's running on x86_64. This is a bug in main opencpn.

There are ways to overcome this for testing, but it's too late now.. Back later. Sweet dreams :)

**Comment by leamas** [comment]
Hm... if you have more time to test this... (please let me know otherwise!)

We have a bug to file against main OpenCPN. But, it should be possible to walk around to see if there are more bugs. The steps:

  - Patch *~/.var/app/org.opencpn.OpenCPN/config/opencpn/ocpn-plugins.xml*, and change on or two `<target>` to *flatpak-aarch64*
    The important is that there is at least one such entry.
  - If not already done: Make yourself to a CatalogExpert by adding the line `CatalogExpert=1`to the [PlugIns] section
    `~/.var/app/org.opencpn.OpenCPN/config/opencpn/opencpn.conf` so it reads

        [PlugIns]
        ...  
        CatalogExpert=1

  - Start opencpn using `-l debug `, and open the *Options | Plugins | Settings* dialog.
  - Here, select flatpak-aarch64 in the *Select New Flavour* dropbox
  - Import the flatpak-aarch64 plugin
  - And submit the log + other findings like "works" or "doesn't work"

**Comment by Rasbats** [comment]
Plenty of time tomorrow. G'Night.

**Comment by leamas** [comment]
The reason the main program consider itself to be flatpak-x86_64 might be that it is a cross-build built on a x86_64 host. Should be possible to patch, but not obvious how to me.

**Comment by Rasbats** [comment]
> And submit the log + other findings like "works" or "doesn't work"

Changed the arch for DR_pi and vfkaps_pi. Both of these could be installed but showed the jigsaw icons. Enabled the plugins. They did not run and when OpenCPN was restarted had gone back to the non-installed list.

Used Custom to import ShipDriver aarch64 and your amd64. Both came up as incompatible.

> might be that it is a cross-build built on a x86_64 host. 

This looks as if it is the problem. CircleCI runner on a local amd64 machine? Probably too deep for me.

[opencpn.log](https://github.com/Rasbats/shipdriver_pi/files/6134383/opencpn.log)

**Comment by leamas** [comment]
Well, it's atcually about the flathub build environment which builds main opencpn...

Did you take the steps I described [above](https://github.com/Rasbats/shipdriver_pi/issues/153#issuecomment-797740483) when generating that log? it looks like it still thinks it's a x86_64.

**Comment by Rasbats** [comment]
All the steps were those you listed. Can GA help with building the aarch64 OpenCPN?

**Comment by leamas** [comment]
I can rebuild that, it's not a problem. I'm on it, stay tuned...

**Comment by leamas** [comment]
Can you trying after installing the test version using flatpak install https://dl.flathub.org/build-repo/40043/org.opencpn.OpenCPN.flatpakref ? 

If there are no progress I'll try to install in a regular Rpi 3 and see if it gives any meaningful results, we need to shorten the loop.

**Comment by Rasbats** [comment]
Did not install directly. Downloaded and install started. Failed with this message:

![Screenshot from 2021-03-13 10-10-01](https://user-images.githubusercontent.com/2537240/111026894-d4266400-83e4-11eb-9e68-1ad63d673f0b.png)

Is there a fix for this?

**Comment by Rasbats** [comment]
SD Card is 128 GB.

**Comment by leamas** [comment]
Hi, sorry for delay. As for the SD card: https://fedoraproject.org/wiki/Architectures/ARM/Raspberry_Pi#Resizing_the_root_partition

**Comment by Rasbats** [comment]
Messed up... I think I downloaded the flatpakref instead of installing from the URI. Now getting 404 not found. Can you check if **https://dl.flathub.org/build-repo/40043/org.opencpn.OpenCPN.flatpakref** is still available with the link to the beta. Thanks.
BTW Got the Pi /root extended to SD Card max size.

**Comment by Rasbats** [comment]
p.s. This was a good article about resizing:
[https://linuxhint.com/install_fedora_33_raspberry_pi4/](https://linuxhint.com/install_fedora_33_raspberry_pi4/)

**Comment by leamas** [comment]
That link was time limited, you will get the same result now using `flatpak update org.opencpn.OpenCPN`

I have been able to install the flatpak'ed opencpn on a regular rPI B with 1 GB RAM (!). It runs, sort of, but there is not enough memory to debug. There are multiple bugs in main opencpn that must be fixed, but I need to nail them down. Expect the beta branch on flathub to be "dynamic" for some time. Back as soon as I have more details.

One interesting experiment you could do which would be very helpful, if you have time, would be to run opencpn in the debugger. The basic steps:

  - Install the sdk: `flatpak install org.freedesktop.Sdk//20.08`
  - Install the debug symbols: `flatpak install org.opencpn.OpenCPN.Debug`
  - Start a shell in the opencpn sandbox: `flatpak run --command=sh --devel  org.opencpn.OpenCPN`
  - In the shell, start the debugger on opencpn: `gdb /app/bin/opencpn`
  - In gdb, set a breakpoint on the isCompatible function: `break isCompatible`
  - in gdb, start opencpn: `run`
  - Using opencpn, start importing a plugin. The breakpoint will trigger and stop program.
  - In gdb, use the 'step' command to advance to the line starting with DEBUG_LOG (around 7 lines)
  - Check some interesting values using '`print compatOS`, `print compatOsVersion` and `print os_detail->osd_arch.c_str()`, and report back

**Comment by leamas** [comment]
Never mind, I messed up (forgot both to invoke prober branch and the -l debug option) .  

Filed new bug: https://github.com/OpenCPN/OpenCPN/issues/2185. These issues must  be sorted out before we can proceed.

**Comment by leamas** [comment]
The opencpn issues are now dealt with in https://github.com/OpenCPN/OpenCPN/pull/2188. I have also rebuilt the flathub beta so it  is built from the same code as the PR.` flatpak update org.opencpn.OpenCPN//beta` should make the trick.

I have also tested the very basic "Import plugin" on a regular rPI of mine with 1MB. It seems to work, but I cannot do anything useful due to memory constraints. If you have time, it would be really nice if you could do some more sane testing with your rPI 4. Otherwise, we should perhaps just close this issue.

**Comment by Rasbats** [comment]
> if you could do some more sane testing with your rPI 4

Just in from the garden... I will have a look later today or tomorrow. Not sure about the sane bit.

**Comment by Rasbats** [comment]
Had a problem with Fedora 33 and the HDMI display. A series of coloured lines. Previous small SD Card was fine.

Moved to Ubuntu 20.10 Desktop 64bit. Pi imager recognised the size of the card so no need to resize.

Tried a number of things, eventually installing the non-working stable OpenCPN. Then tried installing the debug version. No luck.

Removed all the OpenCPN and installed just OpenCPN.debug. Seems the Debug aarch64 version is installed but does not run.

Noticed you used freedesktop 20.04. I seem to be on 18.08.

[opencpn.debug.install.2.txt](https://github.com/Rasbats/shipdriver_pi/files/6154639/opencpn.debug.install.2.txt)

[opencpn.debug.install.txt](https://github.com/Rasbats/shipdriver_pi/files/6154600/opencpn.debug.install.txt)

**Comment by Rasbats** [comment]
Wait... installing freedesktop 20.04.

**Comment by Rasbats** [comment]
No... same error message:

`error: app/org.opencpn.OpenCPN/aarch64/master not installed`

**Comment by leamas** [comment]
You have the wrong version installed, for sure. This is how it looks for me: 

```
name                        Application ID                            Version               Branch       Origin               Installation
Freedesktop Platform        org.freedesktop.Platform                  20.08.6               20.08        flathub              system
default                     org.freedesktop.Platform.GL.default                             20.08        flathub              system
Freedesktop SDK             org.freedesktop.Sdk                       20.08.6               20.08        flathub              system
OpenCPN                     org.opencpn.OpenCPN                       5.2.4-0+dc4c8a2       beta         opencpn-origin       user

```

The basic problem is that you are using the ordinary flathub repo, and thus gets opencpn from the stable branch (I need to fix this so there is no aarch 64 build, BTW). I suggest that you:

  - Removes the current installation (with or without using sudo):  `flatpak uninstall org.opencpn.OpenCPN`
  - Add the beta remote: `flatpak --user remote-add flathub-beta https://flathub.org/beta-repo/flathub-beta.flatpakrepo` 
  - Installs opencpn from the beta branch`: flatpak --user install flathub-beta org.opencpn.OpenCPN//beta`

**Comment by Rasbats** [comment]
The flatpak system is impressive. Good feedback from running the flatpak commands.

[gdb.log](https://github.com/Rasbats/shipdriver_pi/files/6155307/gdb.log)

Forgot to step through.

Opened without gdb... success. Imported ShipDriver. It works.

Went back to gdb.

[gdb.second.test.log](https://github.com/Rasbats/shipdriver_pi/files/6155448/gdb.second.test.log)

Cannot step through the error.

Can I test anything else on this? We now have mirrored systems (well almost).

**Comment by leamas** [comment]
> Opened without gdb... success. Imported ShipDriver. It works.

So, there is no error to debug, after all changes in OpenCPN (?) This seems to verify that the opencpn changes (which I have  made to the flatpak beta channel) fixes the import error.

Now, the question is if the plugin side of this is ok. However, this should be another issue, if any.  Can we actually close this?

**Comment by leamas** [comment]
BTW: Dave merged the changes to main OpenCPN more or less instantly.  Will make another update of the flatpak beta, this time built from opencpn master. Should not affect anything, though.

**Comment by leamas** [comment]
I think we can close this. The basic bits & pieces is in place: the bugs in main openpcn are fixed, and we have built plugins which can verfiy this. 

Closing
